solidity/contracts/bank/Bank.sol

```
@@ -0,0 +1,299 @@
+        uint256 amount,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external {
+        /* solhint-disable-next-line not-rely-on-time */
+        require(deadline >= block.timestamp, "Permission expired");
+
+        // Validate `s` and `v` values for a malleability concern described in EIP2.
+        // Only signatures with `s` value in the lower half of the secp256k1
+        // curve's order and `v` value of 27 or 28 are considered valid.
+        require(
+            uint256(s) <=
+                0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF5D576E7357A4501DDFE92F46681B20A0,
+            "Invalid signature 's' value"
+        );
+        require(v == 27 || v == 28, "Invalid signature 'v' value");
+
+        bytes32 digest =
+            keccak256(
+                abi.encodePacked(
+                    "\x19\x01",
+                    DOMAIN_SEPARATOR(),
+                    keccak256(
+                        abi.encode(
+                            PERMIT_TYPEHASH,
+                            owner,
+                            spender,
+                            amount,
+                            nonce[owner]++,
+                            deadline
+                        )
+                    )
+                )
+            );
+        address recoveredAddress = ecrecover(digest, v, r, s);
+        require(
+            recoveredAddress != address(0) && recoveredAddress == owner,
+            "Invalid signature"
+        );
+        _approveBalance(owner, spender, amount);
+    }
+
+    /// @notice Increases balances of the provided `recipients` by the provided
+    ///         `amounts`. Can only be called by the Bridge.
+    /// @dev This function fails if the lengths of the arrays are not the same.
+    function increaseBalances(
+        address[] calldata recipients,
+        uint256[] calldata amounts
+    ) external onlyBridge {
+        for (uint256 i = 0; i < recipients.length; i++) {
```

Maybe it's worth requiring that `recipients` and `amounts` have the same length to fail fast here? Worth noting that if `recipients` is shorter than `amounts` that function won't revert so the `@dev` comment is not accurate here.

---

Agreed. I was trying to optimize gas as much as possible, but this validation feels crucial and is not that gas expensive. Please see https://github.com/keep-network/tbtc-v2/pull/84/commits/8bd182cc69cb964710a120daf550d3113506a672.